sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -199,23 +199,23 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
     // TimestampWritable.floatToTimestamp
     case FloatType =>
       buildCast[Float](_, f => try {
-        decimalToTimestamp(Decimal(f))
+        decimalToTimestamp(Decimal(f.toString))
```

Comment:
Implicit cast float to double would fail
```scala
val millis = 15 * 1000 + 2
checkEvaluation(
      cast(cast(millis.toFloat / 1000, TimestampType), FloatType),
      millis.toFloat / 1000)
```
due to underline fractional change.